src/core/job.c

```diff
@@ -107,6 +114,13 @@ Job* job_free(Job *j) {
         return mfree(j);
 }
 
+void job_add_triggering_unit(Job *j, Unit *u) {
+        assert(j);
+        assert(u);
+
+        LIST_APPEND(triggered_by, j->triggered_by, u);
```

Comment:
so this is too simple. What if you have three units A, B, C and they have the full set of OnSuccess= deps on each other, i.e. each unit is both the sender and the receiver of each dependency onto two other units? 

This won't work because using a simple linked list means each unit can be part of exactly one job "triggered_by" list, but you don't guard again in anyway against multiple uses.

This cannot use a linked list, frankly. Must be some other structure (Set maybe).

Reply:
Quite frankly, I don't understand why you need this at all though. The reverse deps should already deliver all you need?

Reply:
> This won't work because using a simple linked list means each unit can be part of exactly one job "triggered_by" list, but 
> you don't guard again in anyway against multiple uses.

Sorry, not sure I follow - each job has its own list, so a unit can be referenced in any of them, no?

Reply:
> See this discussion here: https://github.com/systemd/systemd/pull/20833#discussion_r76292286

What precisely do you mean? Why keep another list? I don't get it?

Reply:
> Sorry, not sure I follow - each job has its own list, so a unit can be referenced in any of them, no?

Yeah, but a single unit cannot appear in multiple jobs.

Reply:
> > Sorry, not sure I follow - each job has its own list, so a unit can be referenced in any of them, no?
> 
> Yeah, but a single unit cannot appear in multiple jobs.

Not as the triggere**d**, but it can as the triggere**r**, no?

> > See this discussion here: [#20833 (comment)](https://github.com/systemd/systemd/pull/20833#discussion_r76292286)
> 
> What precisely do you mean? Why keep another list? I don't get it?

As far as I understand, if A and B triggered C, and the jobs get merged because it happens at the same time, you lose the information. By keeping this list, we always retain that information and can report it.